Sync3.23.1

kuchungmsft opened the pull request:
Merge changes from v3.23.1

Only one file "Source/cmVisualStudio10TargetGenerator.cxx" had merge conflicts, I referred to tags v3.22.1 and v3.23.1 to resolve merge conflicts.

Here are the conflicts:

![image](https://user-images.githubusercontent.com/105310954/168931199-a69d4ec3-fde0-4d67-96cd-61db6b83626b.png)

![image](https://user-images.githubusercontent.com/105310954/168931267-e8f902e5-ea87-4293-9a98-ff357e9e988f.png)

![image](https://user-images.githubusercontent.com/105310954/168931323-eb06bd15-1a67-4543-86fa-b7958db523f8.png)

![image](https://user-images.githubusercontent.com/105310954/168931352-b40977d8-5cfb-411f-ac3a-650e17f23caf.png)

![image](https://user-images.githubusercontent.com/105310954/168931397-aaee537c-2809-4538-bc8a-237d9b515ff7.png)

zeule committed:
Help: Fix typo in install command documentation

bradking committed:
Merge topic 'ninja-gfortran-preprocess' into release-3.23

6b4885b58b Ninja: Avoid preprocessing twice with explicit Fortran_PREPROCESS

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7008

friendlyanon committed:
Help: Add missing versionadded to *CROSSCOMPILING_EMULATOR

The list capability was added in CMake 3.15 by commit fec441ec17 (Teach
CROSSCOMPILING_EMULATOR to support arguments, 2019-05-30,
v3.15.0-rc1~6^2), but the documentation did not indicate the version
that added it.

craigscott-crascit committed:
find_file: Fix blank line instead of framework path in debug output

There's no point printing a blank line if FindHeaderInFramework()
returns an empty string. Pass in the debug object to that function
instead so it can record the actual paths it tests in the debug output
directly.

bradking committed:
FindThreads: Revert "Honor THREADS_PREFER_PTHREAD_FLAG when ... in libc"

Revert commit 5efb6fb516 (FindThreads: Honor THREADS_PREFER_PTHREAD_FLAG
when pthread is found in libc, 2021-11-03, v3.21.5~4^2).  The check for
the `-pthread` flag can pass on compilers like XL, that interprets it as
`-p -t hread` and returns zero.  Prior to that commit, we did not use
the check in the `CMAKE_HAVE_LIBC_PTHREAD` code path.  Now we do, it
succeeds, and we incorrectly add the `-pthread` flag for XL.

This change was backported to the 3.21 and 3.22 release series long
after they initially came out.  Since there may be more cases where we
now add `-pthread` incorrectly, it is simplest to revert the change in
all release series pending further investigation.

Fixes: #23270

bradking committed:
Merge branch 'FindThreads-revert-libc-pthread-flag' into release-3.21

Merge-request: !7025

bradking committed:
Merge branch 'FindThreads-revert-libc-pthread-flag' into release-3.22

Merge-request: !7025

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
CUDA: Simplify CMAKE_CUDA_ARCHITECTURES special value logic

Refactor the logic checking `CMAKE_CUDA_ARCHITECTURES` special values.
Switch on the value first, and then make other decisions for each case.
This makes room for other special values to be added later.

bradking committed:
Help: Update CUDA_ARCHITECTURES docs for generic all/all-major support

Since commit 8f64df0a7c (CUDA: Generic all and all-major support,
2021-12-19, v3.23.0-rc1~23^2), these special values do not require any
specific compiler or version.

While at it, re-organize CUDA_ARCHITECTURES special value documentation.
Move the `versionadded` markup inside each special value so we can add
more in future versions.

KyleFromKitware committed:
target_sources(): Prohibit FILE_SET on custom targets

Fixes: #23262

bradking committed:
ci: add cuda11.6 base image

bradking committed:
gitlab-ci: add jobs testing cuda11.6 with nvcc and clang 13

craigscott-crascit committed:
Help: Improve wording and structure related to preset includes

Encourage placing preset includes near the beginning of a preset
file and ensure the example shows that usage. Move the prose
discussing includes to its own section to improve discoverability
and break up paragraphs to make each main point harder to miss.

Also clarify ${sourceDir} to remove any ambiguity with regard to
its meaning in included files.

Issue: #23214

bradking committed:
Merge topic 'FindThreads-revert-libc-pthread-flag' into release-3.23

c6da90bd39 FindThreads: Revert "Honor THREADS_PREFER_PTHREAD_FLAG when ... in libc"

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7025

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'doc-crosscompiling-emulator-arg-list-3.15' into release-3.23

6ff1217b9c Help: Add missing versionadded to *CROSSCOMPILING_EMULATOR

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7023

bradking committed:
CUDA: Fix CMAKE_CUDA_ARCHITECTURES=all/all-major with NVCC 11.5+

Changes in commit 8f64df0a7c (CUDA: Generic all and all-major support,
2021-12-19, v3.23.0-rc1~23^2) broke our architecture verification checks
when using `-arch={all,all-major}` with NVCC 11.5+.  If we test the
compiler with `-arch={all,all-major}`, we have no expected list of
architectures, so skip the check.

Fixes: #23278

robertmaynard committed:
cmake: Correct regression in `-B <dir> <src_dir>`

Fixes: #23285

Joker-Van committed:
KWSys: SystemInformation: Add Zhaoxin x86 processor kh40000

Backport KWSys commit `dfec20c95` (SystemInformation: Add Zhaoxin x86
processor kh40000, 2022-03-02) to the CMake 3.23 release branch.

bradking committed:
Merge topic 'file-set-no-custom-targets' into release-3.23

ad41c9cd11 target_sources(): Prohibit FILE_SET on custom targets

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Merge-request: !7027

bradking committed:
Merge topic 'find_file_frameworks_debug_output' into release-3.23

3354d52e3d find_file: Fix blank line instead of framework path in debug output

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7024

bradking committed:
Merge topic 'doc-presets-v4' into release-3.23

6404751176 Help: Improve wording and structure related to preset includes

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7028

bradking committed:
Merge topic 'cuda-arch-all' into release-3.23

2796d6eeca CUDA: Fix CMAKE_CUDA_ARCHITECTURES=all/all-major with NVCC 11.5+
e450d55552 Help: Update CUDA_ARCHITECTURES docs for generic all/all-major support
fe64c49e72 CUDA: Simplify CMAKE_CUDA_ARCHITECTURES special value logic

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7026

bradking committed:
Tests: Clarify RunCMake.CommandLine ExplicitDirs case names

bradking committed:
Tests: Verify RunCMake.CommandLine ExplicitDirs source/build dirs

bradking committed:
Tests: Add RunCMake.CommandLine ExplicitiDirs explicit work directory

Verify that the work directory is not selected as the source or build
tree.

Issue: #23285

bradking committed:
Tests: Add cases for CMAKE_CUDA_ARCHITECTURES={all,all-major}

The `CudaOnly.All` test only sets these in project code after CUDA is
enabled.  Add another case to test the values during compiler detection.

Issue: #23161

bradking committed:
VS: Fix CUDA compiler id with CMAKE_CUDA_ARCHITECTURES={all,all-major}

Skip the architecture verification check for these values on Visual
Studio.  It cannot be implemented correctly until future work delays the
check to the main compiler test step.

Issue: #23164, #23161

KyleFromKitware committed:
target_sources(): Enforce stricter requirements for FILE_SET name

Fixes: #23286

bradking committed:
Merge topic 'update-kwsys' into release-3.23

a20ad92022 KWSys: SystemInformation: Add Zhaoxin x86 processor kh40000

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7032

bradking committed:
Merge topic 'cmake-dir-args' into release-3.23

d2e9478321 Tests: Add RunCMake.CommandLine ExplicitiDirs explicit work directory
78e8f11456 cmake: Correct regression in `-B <dir> <src_dir>`
4091d5c58c Tests: Verify RunCMake.CommandLine ExplicitDirs source/build dirs
1b97f8f35c Tests: Clarify RunCMake.CommandLine ExplicitDirs case names

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Merge-request: !7030

bradking committed:
Merge topic 'cuda-arch-all' into release-3.23

dae3ad08fa Tests: Add cases for CMAKE_CUDA_ARCHITECTURES={all,all-major}
5c1f5357b0 VS: Fix CUDA compiler id with CMAKE_CUDA_ARCHITECTURES={all,all-major}

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7035

bradking committed:
gitlab-ci: run CUDA and HIP test jobs in any non-MR pipeline

Previously we only ran these in a scheduled nightly pipeline.
We should run them in pipelines on integration branches too,
particularly for the `release` branch.

bradking committed:
BinUtils: Avoid llvm-ar on Apple platforms

Since commit cf82300a63 (BinUtils: Clarify search logic and make it more
consistent, 2021-05-27, v3.21.0-rc1~119^2~2) we correctly prefer the
more-specific name `llvm-ar` over `ar` when using Clang.  However, on
Apple platforms, `llvm-ar` does not generate a symbol table that the
Apple linker accepts.  Fall back to `ar` on Apple platforms.

Fixes: #23269

bradking committed:
Merge branch 'binutils-llvm-ar-clang-macos' into release-3.21

Merge-request: !7039

bradking committed:
Merge branch 'binutils-llvm-ar-clang-macos' into release-3.22

Merge-request: !7039

bradking committed:
Merge branch 'release-3.21' into release-3.22

KyleFromKitware committed:
target_sources(): Process multiple FILE_SET arguments per block

Fixes: #23287

zaufi committed:
while: Restore tolerance of condition error

Since commit 880ca66b51 (Fix: `while()` can silently ignore incorrect
condition, 2021-08-09, v3.22.0-rc1~238^2~4) we correctly reject the
code

    set(paren "(")
    while(${paren})
    endwhile()

However, rejecting it breaks compatibility with projects that used such
code accidentally.  In CMake 3.21 and below, any error in the condition
was ignored because the `false` result exited the loop first.  Restore
tolerance of the error for now.  A policy will be needed to make it an
error later.

Note that the same condition with `if` was always correctly rejected.

Fixes: #22524
Issue: #23296
Co-authored-by: Brad King <brad.king@kitware.com>

bradking committed:
Tests: Simplify RunCMake.{if,while} unbalanced parenthesis cases

bradking committed:
Merge branch 'while-bug-compatibility' into release-3.22

Merge-request: !7041

craigscott-crascit committed:
Help: Update install() docs to better reflect preference for file sets

Previously, headers would typically be installed as bare files or
as whole directories. File sets offer a better abstraction and
associate headers with a target, installing them as part of the
target. Add notes and update examples to draw the reader's
attention to the advantages of file sets for headers.

bradking committed:
Merge topic 'binutils-llvm-ar-clang-macos' into release-3.23

29ea94e17c BinUtils: Avoid llvm-ar on Apple platforms

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7039

bradking committed:
Merge topic 'while-bug-compatibility' into release-3.23

da2361ffb3 while: Restore tolerance of condition error
47d197745a Tests: Simplify RunCMake.{if,while} unbalanced parenthesis cases

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7041

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'ci-cuda-11.6' into release-3.23

1f601753d7 gitlab-ci: add jobs testing cuda11.6 with nvcc and clang 13
7716750062 ci: add cuda11.6 base image
77bdcc949d gitlab-ci: run CUDA and HIP test jobs in any non-MR pipeline

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7038

bradking committed:
Merge topic 'file-set-multiple-names' into release-3.23

8c23ecbd93 target_sources(): Process multiple FILE_SET arguments per block

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7040

bradking committed:
Merge topic 'file-set-name-requirements' into release-3.23

b357d334fc target_sources(): Enforce stricter requirements for FILE_SET name

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Merge-request: !7037

bradking committed:
Help: Document explicitly that 'cxx_std_##' features may not add a flag

This is commonly reported incorrectly as a bug, so call out this
behavior explicitly in the documentation.

bradking committed:
CMake 3.21.6

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
CMake 3.22.3

bradking committed:
Merge branch 'release-3.22' into release-3.23

craigscott-crascit committed:
Help: Reorganise FILE_SETS and related properties

Reduce duplication and make the docs for target_sources() focus
on the functionality rather than the properties it modifies.
The properties are a lower level quantity, so put the relevant
details for them in the property documentation. The target_sources()
command only needs to reference the properties, not reproduce
the property documentation.

Improve the cross-referencing between the HEADER_... property
docs. This helps build the mental picture of how they relate to
each other.

Commit:
FindMatlab: Fix logic error when minimum version is not found

Fix logic added by commit 7fdd5128b1 (FindMatlab: Fix version selection
if a version is given, 2021-07-02, v3.22.0-rc1~66^2).  Ensure that
`_list_index` is always initialized to -1, akin to `list(FIND)` not
finding a match.

Issue: #22377

bradking committed:
Merge branch 'FindMatlab-minver' into release-3.22

Merge-request: !7047

bradking committed:
Tests: Make RunCMake.CMP0125 robust to external filesystem content

Revise the test from commit 08db1341a6 (find_*: ensure consistent
behavior for cache variables, 2021-05-03, v3.21.0-rc1~177^2) to avoid
searching outside the test directories.

Fixes: #23299

KyleFromKitware committed:
CMakePresets.json: Add ${fileDir} macro

Fixes: #23214

Commit:
IAR: Fix detection of modern C++ version

The change in commit cc4da8d13a (IAR/CXX: Fix compatibility with CMP0057
OLD, 2022-01-29, v3.23.0-rc1~46^2) broke the detection of C++ version
because the `IN_LIST` operator cannot work directly on a list but
requires a variable.

bradking committed:
Merge topic 'FindMatlab-minver' into release-3.23

bb2c7b7dc0 FindMatlab: Fix logic error when minimum version is not found

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7047

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'doc-lang-std-features' into release-3.23

fbda79f39d Help: Document explicitly that 'cxx_std_##' features may not add a flag

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: antmak <antmak.pub@gmail.com>
Merge-request: !7043

bradking committed:
BinUtils: Avoid llvm-mt because it is missing 'mt' features we use

Since commit cf82300a63 (BinUtils: Clarify search logic and make it more
consistent, 2021-05-27, v3.21.0-rc1~119^2~2) we correctly prefer the
more-specific name `llvm-mt` over `mt` when using Clang.  However, the
`llvm-mt` tool does not yet support all the flags we need in the
implementation of `vs_link_{exe,dll}`.  Prefer plain `mt` for now.

Fixes: #23305

bradking committed:
Merge branch 'binutils-llvm-mt' into release-3.21

Merge-request: !7054

bradking committed:
Merge branch 'binutils-llvm-mt' into release-3.22

Merge-request: !7054

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
Merge topic 'test-cmp0125' into release-3.23

eca72dab8b Tests: Make RunCMake.CMP0125 robust to external filesystem content

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7049

bradking committed:
Merge topic 'cmake-presets-file-dir' into release-3.23

f54507c2f6 CMakePresets.json: Add ${fileDir} macro

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Acked-by: Gerhard Olsson <gerhard.nospam@gmail.com>
Merge-request: !7051

KyleFromKitware committed:
Tests: Disable CMAKE_SYSTEM_PROGRAM_PATH in IgnorePrefixPath test

Some systems set CMAKE_SYSTEM_PROGRAM_PATH, which pollutes the
environment for this test. Erase it before executing the test to get
a clean environment.

Fixes: #23300

craigscott-crascit committed:
Help: Clarify behavior of search ignore-related variables

Not all the behaviors of CMake variables for ignoring search
locations by find_...() commands were fully documented.
Add the missing effects, clarify the wording and restructure the
way the details are assembled to reduce duplication.

Also improve the cross-referencing to ensure all the related
variables are more discoverable.

Issue: #20878

krasznaa committed:
IRSL: Add paths for Intel oneAPI compilers on Linux

They are stored in a slightly different place with oneAPI than they
used to be in PSXE.

A similar change was made for Windows by commit 956160bb9a (IRSL: Fix
search for Windows redist files with Intel Classic compiler, 2021-09-23,
v3.22.0-rc1~88^2), which left a comment about the locations relative to
the Classic and oneAPI compilers.

Fixes: #23310

bradking committed:
gitlab-ci: Update Windows builds to MSVC 19.31 toolset

bradking committed:
gitlab-ci: Update Windows builds to MSVC 19.31 toolset

robertmaynard committed:
CUDA: Restore support for CMAKE_CUDA_ARCHITECTURES=OFF

Fixes: #23309

bradking committed:
Merge branch 'backport-3.22-ci-msvc-19.31' into ci-msvc-19.31

bradking committed:
Merge topic 'doc-file-sets' into release-3.23

4286b72240 Help: Update install() docs to better reflect preference for file sets
ab1b573f41 Help: Reorganise FILE_SETS and related properties

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7031

bradking committed:
Merge topic 'iar-cmp0057' into release-3.23

bd3e23a605 IAR: Fix detection of modern C++ version

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7053

bradking committed:
Merge topic 'binutils-llvm-mt' into release-3.23

73d0d4e4b5 BinUtils: Avoid llvm-mt because it is missing 'mt' features we use

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7054

bradking committed:
Merge topic 'ignore-prefix-path-test-program-path' into release-3.23

11ea24e590 Tests: Disable CMAKE_SYSTEM_PROGRAM_PATH in IgnorePrefixPath test

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7055

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge branch 'irsl-oneapi-linux' into release-3.22

Merge-request: !7057

bradking committed:
Merge branch 'backport-3.22-ci-msvc-19.31' into release-3.22

Merge-request: !7059

bradking committed:
Merge topic 'ci-msvc-19.31' into release-3.23

f2c303f4c7 Merge branch 'backport-3.22-ci-msvc-19.31' into ci-msvc-19.31
e66de43180 gitlab-ci: Update Windows builds to MSVC 19.31 toolset
fade5b5b8c gitlab-ci: Update Windows builds to MSVC 19.31 toolset

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7059

bradking committed:
Merge topic 'irsl-oneapi-linux' into release-3.23

3e84a06b3a IRSL: Add paths for Intel oneAPI compilers on Linux

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7057

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Tests: Fix testing CMAKE_CUDA_ARCHITECTURES=OFF with Clang

Clang does not automatically choose an architecture supported
by the SDK.  The `OFF` value is meant for users to provide their
own architecture flags, so provide one as part of the test.

bradking committed:
Merge topic 'cuda-arch-off' into release-3.23

7024596a26 Tests: Fix testing CMAKE_CUDA_ARCHITECTURES=OFF with Clang
2a79b647fa CUDA: Restore support for CMAKE_CUDA_ARCHITECTURES=OFF

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7058

bradking committed:
CMake 3.23.0-rc3

Commit:
AppleClang: Add C++20 and C++23 flags

Follow up commit 886e27062b (Clang/MSVC: C++20 final flag, C++23
support, 2021-05-29, v3.20.4~7^2) with support for AppleClang.

bradking committed:
Merge branch 'AppleClang-cxx20-cxx23-flags' into release-3.21

Merge-request: !7063

bradking committed:
Merge branch 'AppleClang-cxx20-cxx23-flags' into release-3.22

Merge-request: !7063

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
ExternalProject: Document that LIST_SEPARATOR works for CMAKE_ARGS too

Since `CMAKE_ARGS` is used to construct the default `CONFIGURE_COMMAND`
for CMake-based external projects, the `LIST_SEPARATOR` option works for
it too.

bradking committed:
Help: Add advice for dealing with semicolons in lists

Issue: #23315

craigscott-crascit committed:
Help: Fix typo and improve readability for LINK_LIBRARIES_ONLY_TARGETS

craigscott-crascit committed:
Merge topic 'doc-LINK_LIBRARIES_ONLY_TARGETS' into release-3.23

95cdce829f Help: Fix typo and improve readability for LINK_LIBRARIES_ONLY_TARGETS

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7068

bradking committed:
Merge topic 'AppleClang-cxx20-cxx23-flags' into release-3.23

683e4e9cab AppleClang: Add C++20 and C++23 flags

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7063

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'doc-ignore-prefix-paths' into release-3.23

5cb0a730c9 Help: Clarify behavior of search ignore-related variables

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7056

KyleFromKitware committed:
Help: Clarify interaction of LANGUAGE and HEADER_FILE_ONLY

bradking committed:
Help: Clarify that ENVIRONMENT test properties take ;-separated lists

Follow up commit 79f64cc773 (Help: ENVIRONMENT_MODIFICATION clarify only
supports singular values, 2022-02-23) to specifically call out the list
format.

Issue: #23315

robertmaynard committed:
Help: Document cmake source and binary dir selection rules

Co-authored-by: Brad King <brad.king@kitware.com>

bradking committed:
BinUtils: Restore llvm-ar fallback on Apple platforms

Since commit 29ea94e17c (BinUtils: Avoid llvm-ar on Apple platforms,
2022-03-03, v3.21.6~1^2) we do not consider `llvm-ar` at all on Apple
platforms.  However, there are existing cross-compiling use cases in
which the toolchain has `llvm-ar` but not `ar`.  Prior to the
re-ordering in commit cf82300a63 (BinUtils: Clarify search logic and
make it more consistent, 2021-05-27, v3.21.0-rc1~119^2~2), we preferred
`ar` and then `llvm-ar`.  Restore the original order for Apple.

Fixes: #23320

bradking committed:
target_link_libraries: Restore LINK_ONLY for multiple static lib dependencies

Since commit c1e812ad4f (target_link_libraries: Improve tolerance of
unquoted generator expressions, 2022-02-15, v3.23.0-rc2~11^2) we
accumulate consecutive non-keyword arguments to recover an unquoted
generator expression as a single entry.  When given multiple consecutive
non-genex library names, the grouping breaks our logic that expects each
entry is either a raw target name or a genex.  Revise the logic to only
accumulate multiple arguments when they end inside a partial genex.

This bug caused `target_link_libraries` to stop wrapping static library
private dependencies in `$<LINK_ONLY:...>` for `INTERFACE_LINK_LIBRARIES`
when multiple consecutive library names are given.  Add a test case
covering that behavior.

Fixes: #23302

bradking committed:
Merge branch 'binutils-llvm-ar-apple' into release-3.21

Merge-request: !7077

bradking committed:
Merge branch 'binutils-llvm-ar-apple' into release-3.22

Merge-request: !7077

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
Merge topic 'binutils-llvm-ar-apple' into release-3.23

fee36b7a78 BinUtils: Restore llvm-ar fallback on Apple platforms

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7077

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'language-header-file-only-doc' into release-3.23

83a80e5193 Help: Clarify interaction of LANGUAGE and HEADER_FILE_ONLY

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7075

bradking committed:
Merge topic 'doc-list-semicolons' into release-3.23

8abd714176 Help: Clarify that ENVIRONMENT test properties take ;-separated lists
02cf404ace Help: Add advice for dealing with semicolons in lists
c4117d9116 ExternalProject: Document that LIST_SEPARATOR works for CMAKE_ARGS too

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !7066

dg0yt committed:
FindPostgreSQL: Update known versions in documentation

MarkaPola committed:
FindPython: fix various problems using pypy interpreters

* add more possible directories for include file search
* enhance version detection from library and include files
* search for file pypy_decl.h when PyPy.h is not defined

bradking committed:
Merge topic 'tll-genex-concat' into release-3.23

add64399c5 target_link_libraries: Restore LINK_ONLY for multiple static lib dependencies

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Merge-request: !7078

bradking committed:
Merge branch 'FindPython-pypy-specify-versions' into release-3.22

Merge-request: !7081

bradking committed:
Merge topic 'doc-src-bld-selection' into release-3.23

fea270d3ed Help: Document cmake source and binary dir selection rules

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7076

bradking committed:
Merge topic 'FindPostgreSQL-doc-versions' into release-3.23

fcc0119af0 FindPostgreSQL: Update known versions in documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7080

robertmaynard committed:
cmake: Always prefer the last source directory provided

Fixes: #23334

Commit:
FindMatlab: Add R2022a => 9.12 version

bradking committed:
Merge topic 'always_prefer_last_source_dir' into release-3.23

b44ad7992a cmake: Always prefer the last source directory provided

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7084

bradking committed:
Merge topic 'FindPython-pypy-specify-versions' into release-3.23

2834cea155 FindPython: fix various problems using pypy interpreters

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7081

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Help: Document warning about multiple source paths

Although passing multiple source paths was never documented, it was not
diagnosed by CMake 3.22 and below.  In CMake 3.23 we now diagnose extra
paths and warn.  Document this change and add a release note.

Issue: #23334

bradking committed:
Merge topic 'r2022a' into release-3.23

6032213456 FindMatlab: Add R2022a => 9.12 version

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7086

bradking committed:
Merge topic 'doc-multiple-src-paths' into release-3.23

1b8669c1d6 Help: Document warning about multiple source paths

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7082

bradking committed:
CMake 3.23.0-rc4

bradking committed:
Utilities/Release: Add "source" stage to Windows docker spec

Revise the spec added by commit ff929badb3 (Utilities/Release: Add
docker specs to build and test Windows binaries, 2020-05-05,
v3.18.0-rc1~203^2~1) to add a `source` stage that stops just after
copying the source tree into the image.  This provides more granular
control to driving scripts.

craigscott-crascit committed:
Help: Update the main purpose of define_property()

The BRIEF_DOCS and FULL_DOCS are remnants from before the
Sphinx-based documentation when CMake's internal structures
for defining properties included fields for their documentation.
They are no longer mandatory for define_property() and haven't
been in practical use for some time. The main use of the command
has evolved to now be more about how to initialize and inherit
properties, so update the docs to reflect that change in focus.

Issue: #20698

craigscott-crascit committed:
define_property(): Only test prefix if INITIALIZE_FROM_VARIABLE is given

craigscott-crascit committed:
Help: Clean up INITIALIZE_FROM_VARIABLE define_property() option

INITIALIZE_FROM_VARIABLE is new in CMake 3.23, but the
versionadded note was missing in the original commit. The docs
also failed to mention that the new option only applies to target
properties.

Amends fce24e4f10 (define_property(): Add INITIALIZE_FROM_VARIABLE
argument, 2022-01-13)

craigscott-crascit committed:
Help: Add more detail for CPACK_PRODUCTBUILD_DOMAINS* variables

craigscott-crascit committed:
productbuild: Simplify internal CPACK_PRODUCTBUILD_DOMAINS usage

There's no benefit to storing the CPACK_PRODUCTBUILD_DOMAINS
prefix in a variable and appending to it in the C++ code. It has the
disadvantage of making it harder to find usages of the variables with
a suffix appended to that string. Expand out the strings at the places
they are used so that they are easier to spot.

craigscott-crascit committed:
CPack: Avoid space / tab mix in productbuild distribution.xml template

craigscott-crascit committed:
productbuild: Don't write rootVolumeOnly attribute if writing domains

The rootVolumeOnly attribute is deprecated. Apple docs say to use
domains instead.

Fixes: #23343

robertmaynard committed:
FindCUDAToolkit: Add missing cufftw_static target

Fixes #23347

mclegrand committed:
CPack/NSIS: Fix uninstall command when run from installer

The quoting introduced by commit eb3b3bacdc (CPack/NSIS: Fix uninstall
on Windows using "Apps & Features", 2021-09-13, v3.22.0-rc1~136^2)
created two errors in the uninstaller call: double quoting of the
uninstaller executable, and quotes added to the `_?=` argument which
does not support them.  Simplify the command.

robertmaynard committed:
Help: Better cross-reference CMAKE_FIND_NO_INSTALL_PREFIX

bradking committed:
Revert INTERFACE_LINK_LIBRARIES_DIRECT feature for 3.23 branch

Revert commit f3ad061858 (Add usage requirements to update direct link
dependencies, 2022-01-12, v3.23.0-rc1~44^2) and the property storage
updates in its predecessor commit 193a999cd5 (cmTarget: Add
INTERFACE_LINK_LIBRARIES_DIRECT{,_EXCLUDE} backtrace storage,
2022-01-06, v3.23.0-rc1~44^2~1) from the 3.23 release branch.

After initial experience using the feature in practice, additional
design considerations have been raised for discussion in the original
issue.  To avoid rushing this for the 3.23 series, we've decided to
revert the feature for now so it can be revised for a future release.

Issue: #22496

bradking committed:
Merge topic 'rel-win-docker' into release-3.23

a6aa88e51f Utilities/Release: Add "source" stage to Windows docker spec

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7087

bradking committed:
Merge branch 'rel-win-docker' into release-3.21

Merge-request: !7087

bradking committed:
Merge branch 'rel-win-docker' into release-3.22

Merge-request: !7087

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
Merge branch 'release-3.22' into release-3.23

mclegrand committed:
CPack/NSIS: Add support for unquoted (legacy) uninstaller strings

bradking committed:
Merge branch 'cpack-nsis-fix-uninstall-quoting' into release-3.22

Merge-request: !7096

bradking committed:
Help: Clarify when add_test and test properties support generator expressions

Issue: #23329

osorensen committed:
Tutorial: Fix casing of reference to License.txt

Correct a tiny typo in the Step 7 tutorial.  Change so `License.txt`
correctly starts with a capital `L`.

KyleFromKitware committed:
define_property(): Change constraints of INITIALIZE_FROM_VARIABLE

Remove the requirement that the variable name have a prefix while
keeping the suffix requirement. Require that the property name
contains an underscore. Update docs and tests accordingly.

Fixes: #23340

bradking committed:
Merge topic 'doc-productbuild' into release-3.23

bf44efadab Help: Add more detail for CPACK_PRODUCTBUILD_DOMAINS* variables

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7091

bradking committed:
Merge topic 'FindCUDAToolkit-cufftw-static' into release-3.23

3e7113e644 FindCUDAToolkit: Add missing cufftw_static target

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7093

bradking committed:
Merge topic 'cpack-nsis-fix-uninstall-quoting' into release-3.23

5d2ceaada8 CPack/NSIS: Add support for unquoted (legacy) uninstaller strings
b795c96727 CPack/NSIS: Fix uninstall command when run from installer

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7096

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'productbuild-rootVolumeOnly' into release-3.23

2a8df7e7db productbuild: Don't write rootVolumeOnly attribute if writing domains
95eb8cbcfc CPack: Avoid space / tab mix in productbuild distribution.xml template
28fdc3a536 productbuild: Simplify internal CPACK_PRODUCTBUILD_DOMAINS usage

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7092

bradking committed:
Merge topic 'doc-test-property-genex' into release-3.23

78e0204e81 Help: Clarify when add_test and test properties support generator expressions

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7101

bradking committed:
Merge topic 'cleanup-define_property' into release-3.23

87c3b5e421 define_property(): Only test prefix if INITIALIZE_FROM_VARIABLE is given
9b50f221f6 Help: Update the main purpose of define_property()
e993e2c52c Help: Clean up INITIALIZE_FROM_VARIABLE define_property() option

Acked-by: Kitware Robot <kwrobot@kitware.com>
Tested-by: buildbot <buildbot@kitware.com>
Acked-by: Kyle Edwards <kyle.edwards@kitware.com>
Merge-request: !7090

bradking committed:
Merge topic 'tutorial-license-name' into release-3.23

6892002fcb Tutorial: Fix casing of reference to License.txt

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7100

bradking committed:
Merge topic 'revert-link-interface-direct-for-3.23' into release-3.23

687a91967f Revert INTERFACE_LINK_LIBRARIES_DIRECT feature for 3.23 branch

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7098

bradking committed:
Merge topic 'initialize-property-name' into release-3.23

26caa97057 define_property(): Change constraints of INITIALIZE_FROM_VARIABLE

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7103

bradking committed:
Merge topic 'better_cross-ref_CMAKE_FIND_NO_INSTALL_PREFIX' into release-3.23

7dc654a017 Help: Better cross-reference CMAKE_FIND_NO_INSTALL_PREFIX

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7097

bradking committed:
CMake 3.23.0-rc5

MehdiChinoune committed:
GetPrerequisties: pass "-a" only to grep.

moyo1997 committed:
bindexplib: Exclude entry and exit thunks for arm64EC

craigscott-crascit committed:
CPack/IFW: Add support for QtIFW 4.3

craigscott-crascit committed:
CPack/IFW: Add missing referenced source file validation

craigscott-crascit committed:
Help: Clean up and expand CPack/IFW docs

bradking committed:
Merge topic 'GetPrerequisties-win-findstr' into release-3.23

ef2fa9b03b GetPrerequisties: pass "-a" only to grep.

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7110

bradking committed:
AIX: Fix executable ENABLE_EXPORTS in Makefile generators

In commit afcd9fe669 (AIX: Add an option to disable automatic exports
from shared libraries, 2020-01-30, v3.17.0-rc1~47^2) the population of
the `<AIX_EXPORTS>` rule variable placeholder was accidentally added to
the device linking rule rather than the main linking rule.  This caused
our `ExportImportList` script on AIX, when called for executables with
`ENABLE_EXPORTS` set, to be given an `AIX_EXPORTS` file name that does
not exist, leading to a warning from the `dump` tool.

Move the population of the `<AIX_EXPORTS>` placeholder in the Makefile
generators to the main link rule.

Issue: #20290

bradking committed:
Merge branch 'aix-exe-exports' into release-3.21

Merge-request: !7120

bradking committed:
Merge branch 'aix-exe-exports' into release-3.22

Merge-request: !7120

bradking committed:
Merge branch 'release-3.21' into release-3.22

bradking committed:
Merge topic 'aix-exe-exports' into release-3.23

a8dae21ac1 AIX: Fix executable ENABLE_EXPORTS in Makefile generators

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7120

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'bindexplib-arm64ec' into release-3.23

b019ee1a8d bindexplib: Exclude entry and exit thunks for arm64EC

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7112

bradking committed:
Merge topic 'cpack-ifw-4.3' into release-3.23

343ed1f913 CPack/IFW: Add support for QtIFW 4.3

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7114

bradking committed:
Merge topic 'doc-cpack-ifw' into release-3.23

db774e9108 Help: Clean up and expand CPack/IFW docs

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7115

bradking committed:
Merge topic 'cpack-ifw-validate-input' into release-3.23

e41f1ef6ff CPack/IFW: Add missing referenced source file validation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Acked-by: Konstantin Podsvirov <konstantin@podsvirov.pro>
Merge-request: !7116

bradking committed:
CMake 3.23.0

mathstuf committed:
CMakeDependentOption: improve documentation

This makes it far clearer that `<depends>` is a list up front instead of
burying the lede because a list is generally "trivially true" in CMake[1].
Also clarify that `<force>` is only available as a local variable and if
queried outside of the "scope" of the `cmake_dependent_option` call,
will get the stored user cache value.

[1] The exception being when the last entry ends in `-NOTFOUND`.

Suggested-by: Rui Oliveira

tniessen committed:
Help: Add that CMAKE_CACHEFILE_DIR might not be defined

CMake only sets `CMAKE_CACHEFILE_DIR` when writing `CMakeCache.txt`,
so the variable will usually be undefined when `CMakeLists.txt` runs.
Revise its documentation to clarify that `CMAKE_BINARY_DIR` should
be used instead.

tniessen committed:
Help: Fix string(JSON ... LENGTH ...) signature

The list of indices and/or member names is optional, i.e., can be
empty. For example, string(JSON foo LENGTH "[0,0,0]") sets foo to 3.

bradking committed:
GNUInstallDirs: Revert "Apply Debian multiarch LIBDIR to more prefixes"

Revert commit dd9584b352 (GNUInstallDirs: Apply Debian multiarch LIBDIR
to more prefixes, 2021-11-19, v3.23.0-rc1~323^2).  There are separate
problems with activating multiarch `LIBDIR` for each prefix it added:

* Prefix `/` is often used to stage an installation with `DESTDIR`
  for inclusion in a tarball package or similar.

* Prefix `/usr/local` is the default `CMAKE_INSTALL_PREFIX`, causing
  the multiarch `LIBDIR` to be cached after the first configuration,
  even if the prefix changes later.

Revert the change for now, except for the documentation update.
Further discussion will be needed to select a way to enable
multiarch `LIBDIR` for `/` and `/usr/local`.

Fixes: #23365
Issue: #19698

robertmaynard committed:
CUDA: Ninja generator generates valid compile database

The ninja generator has been updated to poperly write the compile
database when the CUDA language has been enabled.

Fixes #23368

bradking committed:
Merge topic 'CMakeDependentOption-doc-improvement' into release-3.23

7a614977f9 CMakeDependentOption: improve documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Rui Oliveira <ruimail24@gmail.com>
Merge-request: !7124

bradking committed:
Merge topic 'ninja_compile_database_understand_new_cuda_lang_variables' into release-3.23

f7c6b9833c CUDA: Ninja generator generates valid compile database

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7129

bradking committed:
Merge topic 'help-json-length-empty-index' into release-3.23

2f0a3d43e1 Help: Fix string(JSON ... LENGTH ...) signature

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7126

bradking committed:
Merge topic 'help-variables-cmakecachefiledir-might-not-be-defined' into release-3.23

cf4100d7da Help: Add that CMAKE_CACHEFILE_DIR might not be defined

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7125

bradking committed:
Merge topic 'GNUInstallDirs-revert-multiarch-prefixes' into release-3.23

624ec3692f GNUInstallDirs: Revert "Apply Debian multiarch LIBDIR to more prefixes"

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7127

bradking committed:
Help: Clarify when policy CMP0112 takes effect

Fixes: #23381

bradking committed:
Utilities/Sphinx: Add undocumented option to build docs for cmake.org

bradking committed:
Utilities/Sphinx: Add option to build outdated version banner for cmake.org

This was previously added in a custom branch for building the
cmake.org reference documentation.

bradking committed:
Utilities/Sphinx: Add OpenSearch link to html page headers on cmake.org

Reference an OpenSearch description file that sits outside the content
of any specific version so we only need to maintain one and so that
it can reference the latest version.

This was previously added in a custom branch for building the
cmake.org reference documentation.

gjasny committed:
Tests: Ignore all classes in Xcode internal objc warnings

Xcode on Apple Silicon warns not only about AMSupportURL classes
but also many more.

craigscott-crascit committed:
Merge topic 'doc-install-typo' into release-3.23

7431759ebc Help: Fix typo in install command documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7140

bradking committed:
Merge topic 'doc-CMP0112' into release-3.23

9993774172 Help: Clarify when policy CMP0112 takes effect

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7134

bradking committed:
gitlab-ci: consolidate jobs for cmake.org/cmake/help/git-{master,stage} docs

bradking committed:
gitlab-ci: distinguish release and development pipeline schedules

Redefine the `CMAKE_CI_PACKAGE` pipeline schedule variable to
indicate whether it is for a development version (`dev`) or a
release version (`v[0-9]...`).  Use this to automatically turn
package upload jobs on or off without having to edit the jobs
in `.gitlab-ci.yml` for release branches.

bradking committed:
gitlab-ci: clarify name of package upload job template

The template is used to upload both source and binary packages.

bradking committed:
Merge branch 'backport-3.22-ci-package-uploads' into ci-package-uploads

bradking committed:
Merge branch 'backport-3.22-ci-package-uploads' into release-3.22

Merge-request: !7142

bradking committed:
Merge topic 'ci-package-uploads' into release-3.23

b20a19fca1 Merge branch 'backport-3.22-ci-package-uploads' into ci-package-uploads
cb44e0d47c gitlab-ci: distinguish release and development pipeline schedules
3a9a9a3ace gitlab-ci: clarify name of package upload job template

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7142

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
gitlab-ci: update cmake.org documentation in release package pipeline

bradking committed:
gitlab-ci: add sanity check to upload jobs

Ensure that the upload destination is not empty.

bradking committed:
gitlab-ci: simplify package pipeline job conditions

Use the job stage to distinguish upload jobs instead of an
explicit variable.

bradking committed:
gitlab-ci: start release package pipelines manually

A release pipeline is always created by manual execution of a pipeline
schedule.  Require the initial pipeline jobs to be started manually too
so that we can later add separate components to play separately.

bradking committed:
Merge branch 'ci-release-manual' into release-3.22

Merge-request: !7143

bradking committed:
Merge topic 'ci-release-manual' into release-3.23

eb410615f2 gitlab-ci: start release package pipelines manually
3a90800a9c gitlab-ci: simplify package pipeline job conditions
9a1b301c85 gitlab-ci: add sanity check to upload jobs

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7143

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'ci-cmake.org-help' into release-3.23

2808281730 gitlab-ci: update cmake.org documentation in release package pipeline
ed00a29cce gitlab-ci: consolidate jobs for cmake.org/cmake/help/git-{master,stage} docs
5c2e8ce515 Utilities/Sphinx: Add OpenSearch link to html page headers on cmake.org
a14905d4df Utilities/Sphinx: Add option to build outdated version banner for cmake.org
cca73b54ae Utilities/Sphinx: Add undocumented option to build docs for cmake.org

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7144

bradking committed:
Merge branch 'ci-cmake.org-help' into release-3.22

Merge-request: !7144

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
gitlab-ci: Add objects.inv to cmake.org html documentation

bradking committed:
gitlab-ci: Build qthelp-format release documentation for cmake.org

Previously the qthelp-format release documentation on `cmake.org` was
built manually.

bradking committed:
Merge branch 'ci-cmake.org-help' into release-3.22

Merge-request: !7146

bradking committed:
Merge topic 'ci-cmake.org-help' into release-3.23

b691906d27 gitlab-ci: Build qthelp-format release documentation for cmake.org
1ceec19c20 gitlab-ci: Add objects.inv to cmake.org html documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7146

bradking committed:
Merge branch 'release-3.22' into release-3.23

KyleFromKitware committed:
FILE_SET: Forbid adding header sets to Apple FRAMEWORK libraries

The feature needs a specialized implementation to place headers
in the right place inside frameworks.  To avoid silently doing
the wrong thing, make this case an error for the 3.23 series.

Issue: #23386

bradking committed:
Help: Add "Updates" section header in 3.22 release notes

This was accidentally left out in commit 1461eff899 (Help: Add missing
3.22.1 section to the release notes, 2022-01-24, v3.22.2~1^2).

bradking committed:
Utilities/Sphinx: Update qthelp generation to qhelpgenerator

`qcollectiongenerator` is deprecated in favor of `qhelpgenerator`.

bradking committed:
Merge branch 'doc-qhelpgenerator' into release-3.22

Merge-request: !7150

bradking committed:
Merge topic 'header-sets-no-framework' into release-3.23

f779f8c0ad FILE_SET: Forbid adding header sets to Apple FRAMEWORK libraries

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Kyle Edwards <kyle.edwards@kitware.com>
Merge-request: !7149

bradking committed:
CheckCompilerFlag: Fix regression in locale environment preservation

Fix a typo from commit 660e0d80ae (internal/CheckCompilerFlag: rely on
common configuration, 2022-01-12, v3.23.0-rc1~124^2~1) that caused
locale environment variables to not be restored after they are set
during the check.

mathstuf committed:
cmFileSet: store visibility with the fileset

The visibility is intrinsic to the fileset, so store it with it. This
avoids recalculating it on every addition to the fileset.

bradking committed:
Merge branch 'doc-qhelpgenerator' into release-3.23

Merge-request: !7150

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'CheckCompilerFlag-restore-env' into release-3.23

9dd585d12c CheckCompilerFlag: Fix regression in locale environment preservation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7153

craigscott-crascit committed:
Help: Correct/add missing details for CheckPIESupported

craigscott-crascit committed:
CheckPIESupported: Prevent separate output lines running together

craigscott-crascit committed:
CheckSourceCompiles: Avoid linker warning with -fembed-bitcode

When the Apple linker sees -headerpad_max_install_names and
bitcode is enabled with a flag like -fembed-bitcode, it issues a warning
and ignores the -headerpad_max_install_names flag. This causes
unrelated compiler and linker flag checks to fail for valid flags.
In f745e0497e (CheckCompilerFlags: Catch linker warning about ignored
flags, 2022-01-03), we started detecting linker warnings, which caused
a regression for projects that were setting -fembed-bitcode in their
CMAKE_CXX_FLAGS or similar. Prevent that regression by removing
the -headerpad_max_install_names linker flag when we know it will
warn and be ignored anyway.

Fixes: #23390
Issue: #23408

rouault committed:
FindPostgreSQL: add support for PostgreSQL 14

bradking committed:
Merge topic 'check-flag-fembed-bitcode-conflict' into release-3.23

a10fc754a6 CheckSourceCompiles: Avoid linker warning with -fembed-bitcode

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7156

robertmaynard committed:
cmake: --list=presets=[type] doesn't generate incorrect warnings

Fixes #23407

cristianadam committed:
PCH: Fix Xcode non-pch language exclusion

Fix a regression caused by commit bbcdac4e5d (PCH: Fix all-language
precompile header support in Xcode, 2021-08-07, v3.22.0-rc1~140^2).

Fixes: #23138

mathstuf committed:
cmTarget: make HEADER_SETS and INTERFACE_HEADER_SETS read-only

There is no reason to allow these properties to be manipulated by user
code. Instead, use the stored visibility on the fileset objects to
derive what these properties should contain.

bradking committed:
Merge topic 'xcode-pch' into release-3.23

476c6a8910 PCH: Fix Xcode non-pch language exclusion
fcf1fcfd0c Tests: Ignore all classes in Xcode internal objc warnings

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7166

bradking committed:
Merge branch 'xcode-pch' into release-3.22

Merge-request: !7166

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'correct_list-presets_parsing' into release-3.23

45299a8f9b cmake: --list=presets=[type] doesn't generate incorrect warnings

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: buildbot <buildbot@kitware.com>
Merge-request: !7164

bradking committed:
Merge topic 'FindPostgreSQL-14' into release-3.23

23671318f3 FindPostgreSQL: add support for PostgreSQL 14

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7158

bradking committed:
Merge topic 'doc-CheckPIESupported' into release-3.23

91c859d1e5 CheckPIESupported: Prevent separate output lines running together
ab5455aa79 Help: Correct/add missing details for CheckPIESupported

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7157

bradking committed:
Help: Document that target_sources defines [INTERFACE_]HEADER_SETS

bradking committed:
CMake 3.22.4

bradking committed:
Merge branch 'release-3.22' into release-3.23

bradking committed:
Merge topic 'file-set-repr-improvements' into release-3.23

5fa15ec9f3 Help: Document that target_sources defines [INTERFACE_]HEADER_SETS
c5d4812f20 cmTarget: make HEADER_SETS and INTERFACE_HEADER_SETS read-only
05783b168d cmFileSet: store visibility with the fileset

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !7168

bradking committed:
CMake 3.23.1

kuchungmsft committed:
Merge tag 'v3.23.1' into dev/kuchung/sync3.23.1

CMake 3.23.1

bobbrow commented:
I did some manual diff'ing of files against the upstream copy and it looks ok to me.

kuchungmsft committed:
Fix Unit Test Failures Due to Microsoft_CMake_VERSION_PATCH Contaminates Tweak Field

nasinlock reviewed (approved):
I took a quick look at the diffs and nothing stood out. It's hard to review the whole thing. It would be good if there was some other information showing that the code changes resulted in CMake 3.23.1 being produced from the code base.

kuchungmsft committed:
Fix Unit Test Failure on Build Agent

Our pipelines set path to somehting like 202205161, which would be greater than minimum required patch.
We want this test case to fail, we need to increase minimum required patch.
